docs/src/components/Header.astro

```diff
@@ -0,0 +1,334 @@
+  const search = document.getElementById('search');
+  const searchBar = document.getElementById('search-bar');
+  const searchInput = document.getElementById('search-input') as HTMLInputElement;
+  const searchStats = document.getElementById('search-stats');
+  const searchResults = document.getElementById('search-results');
+  let searchData: [{ text: string; route: string; content: string }] | null = null;
+  let lastScroll = 0;
+
+  slot!.addEventListener('scroll', () => {
+    if (slot!.scrollTop > 0 && searchBar!.style.display == 'none') {
+      header!.style.borderBottomWidth = '1px';
+    } else {
+      header!.style.borderBottomWidth = '0px';
+    }
+    if (Math.abs(lastScroll - slot!.scrollTop) > 50) {
+      if (
+        window.innerWidth < 1000 &&
+        slot!.scrollTop > lastScroll &&
+        !header!.classList.contains('header-hidden')
+      ) {
+        header!.classList.add('header-hidden');
+      } else if (
+        window.innerWidth < 1000 &&
+        slot!.scrollTop < lastScroll &&
+        header!.classList.contains('header-hidden')
+      ) {
+        header!.classList.remove('header-hidden');
+      }
+      lastScroll = slot!.scrollTop;
+    }
+  });
+
+  search!.addEventListener('click', () => {
+    if (!searchData) {
+      fetch('/search.json')
+        .then((r) => r.json())
+        .then((r) => (searchData = r));
+    }
+    if (searchBar!.style.display == 'none') {
+      searchBar!.style.display = 'block';
+      header!.style.borderBottomWidth = '0px';
+      searchInput!.focus();
+    } else {
+      searchBar!.style.display = 'none';
+    }
+  });
+
+  document.addEventListener('keydown', (e) => {
+    if (e.key == '/') {
+      if (!searchData) {
+        fetch('/search.json')
+          .then((r) => r.json())
+          .then((r) => (searchData = r));
+      }
+      if (searchBar!.style.display == 'none') {
+        searchBar!.style.display = 'block';
+        header!.style.borderBottomWidth = '0px';
+        searchInput!.focus();
+      } else {
+        searchBar!.style.display = 'none';
+      }
+      e.preventDefault();
+    }
+  });
+
+  searchInput!.addEventListener('input', () => {
+    if (searchData) {
+      if (!searchInput.value.length) {
+        searchResults!.replaceChildren();
+        searchStats!.style.display = 'none';
+        return;
+      }
+      searchStats!.style.display = 'inline-block';
+      searchResults!.replaceChildren(
+        ...searchData
+          .filter(
+            (s) =>
+              s.content.toLowerCase().includes(searchInput.value.toLowerCase()) ||
+              s.text.toLowerCase().includes(searchInput.value.toLowerCase())
+          )
+          .map((s) => {
+            let anchor = document.createElement('a');
+            anchor.classList.add('search-result');
+            anchor.href = s.route;
+            let header = document.createElement('h3');
+            header.innerHTML = s.text.replace(
```

Comment:
xss moment :trollface: /j

Reply:
> xss moment :trollface: /j

I think we'd have bigger issues if XSS got onto a docs page